# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def addTwo(x):
    return x + 2

print(addTwo(4))

# Lambda function syntax
a_var = 4
print((lambda x: x+2)(a_var))

6
6


#### Saving to a Variable

In [1]:
f_test = lambda x: x+2
f_test(4)

6

#### Multiple Inputs

In [4]:
# Multiple inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

x_test = lambda x,y,z: x*y*z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [5]:
def multiply(f, num):
    """
    f expects a lambda function
    num expects a number
    """
    return f(num)
multiply(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [11]:
def multiply_test(num):
    return num*4

# Function in a function

def returnFunc():
    test = 4
    def multiply(num):
        return num*2
    return multiply

f_return = returnFunc()
print(f_return)
print(f_return(4))

# Lambda function return from a regular function
def returnLam(b,c):
    return lambda x,a: x+a+b+c
r_lamb = returnLam(4,5)
print(r_lamb)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x000001F91F0FE710>
8
<function returnLam.<locals>.<lambda> at 0x000001F91F291360>
19


#### If Statements within Lambdas

In [13]:
# Lambda x: True if (condition) else False
f_condition = lambda num:num *2 if num >10 else num+2

print(f_condition(8))
print(f_condition(20))

10
40


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [3]:
f = lambda x: x ** 3
print (f(2))

8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [14]:
"""
map(function,iterable)
an iterable is something you can loop over; such as a list, dict, tuple, set orstring
Normally, the usage of map happens with a pre-defined function, but we can use lambdas as well
"""
def squared(num, num2):
    if num <10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
numbers = [4,11,20,3,15,20]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [15]:
"""
map(lambda x: do something, iterable)
using lambda in map usually all takes place on one line
"""
squared_nums_lamb = list(map(lambda x,y: (x**2, y**2) if x <10 and y<10 else (x,y), numbers, more_nums))
print(squared_nums_lamb)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [7]:
num_list = [2, 8, 12, 17]

double_minus_one = list(map(lambda x: (x * 2)- 1, num_list))

print(double_minus_one)


[3, 15, 23, 33]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [6]:
# list(filter(function, iterable))
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else: 
        return False

new_names_list = list(filter(a_names,names))
print(new_names_list)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [7]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [18]:
#1 way
import statistics

num_list = [4, 8, 16, 20, 32, 2, 97]
filtered_nums = list(filter(lambda x: x < mean, num_list))
print(filtered_nums)

# another way

import numpy as np
import statistics
print(sum(numbers)/len(numbers))
print(list(filter(lambda x: True if x >= sum(numbers)/len(numbers) else False, numbers)))
print(list(filter(lambda x: True if x >= np.mean(numbers else False, numbers)))
print(list(filter(lambda x: True if x >= statistics.mean(numbers) else False, numbers)))

[4, 8, 16, 20, 2]


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [2]:
def add_nums(num):
    # Set a base case to stop recursion from going for forever!!!
    if num <= 1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
    return num + add_nums(num-1)

add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [4]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num-1)

factorial(5)

120

#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [2]:
def sub(num1, num2):
    #Base case
    if num2 == 0 :
        return num1
    else:
        print(num1, num2)
        return sub(num1-1, num2-1)

sub(25,5)

TypeError: sub() missing 1 required positional argument: 'num'

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [6]:
def my_range(stop,start=0, step=2):
    while start < stop:
        yield start #yield means generator, it won't return anything
        start += step
        
for i in my_range(10):
    print(i)
print([i for i in my_range(10)])

0
1
2
3
4
0
2
4
6
8
[0, 2, 4, 6, 8]


#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [13]:
def squared_nums(num):
    while num > 0:
        yield num**2
        num -=1

[i for i in squared_nums(5)]

[25, 16, 9, 4, 1]

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [18]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filtered_list = filter(lambda x: x.strip() != '', places)
new_list = list(filtered_list)

print(new_list)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [24]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda x: x.split()[-1].lower())

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [32]:
# F = (9/5)*C + 32
places_celcius = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
places_fahrenheit = list(map(lambda places_temp: (places_temp[0], (9/5) * places_temp[1] + 32), places_celcius))
print(places_fahrenheit)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [34]:
def fib(n, iteration=0, result=None):
    if result is None:
        result = []

    if iteration <= n:
        if iteration == 0 or iteration == 1:
            value = 1
        else:
            value = result[-1] + result[-2]

        result.append(value)
        print(f'Iteration {iteration}: {value}')
        fib(n, iteration + 1, result)

fib(4)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
